docs/spec_draft.md

```diff
@@ -483,6 +484,59 @@ IEEE-754 specification.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.cholesky
+
+### Semantics
+
+Computes the Cholesky decomposition of a batch of symmetric (Hermitian) positive
+definite matrices.
+
+If `lower` is `true`, it computes the lower-triangular matrices `l` such that
+$a = l \cdot l^{T}$. If `lower` is `false`, it computes the upper-triangular
+matrices `u` such that $a = u^{T} \cdot u$.
+
+If the rank of `a` is greater than 2, all except the minor 2 dimensions of `a`
+are treated as batch dimensions.
+
+If `a` is not symmetric (Hermitian) positive definite, the result is
+implementation-defined.
```

Comment:
How about replacing the Semantics section as follows:

```
Computes the Cholesky decomposition of a batch of matrices.

More formally, for all `i`, `result[i0, ..., iR-3, :, :]` is a Cholesky 
decomposition of `a[i0, ..., iR-3, :, :]`, in the form of either of a 
lower-triangular (if `lower` is `true`) or upper-triangular (if `lower` is
`false`) matrix. The output values in the opposite triangle, i.e. the strict
upper triangle or strict lower triangle correspondingly, are 
implementation-defined.

If there exists `i` where the input matrix is not an Hermitian
positive-definite matrix, then the result is implementation-defined.
```

Feedback:
  1. Cholesky decomposition already implies Hermitian, positive-definite matrices, so there's no need to repeat that in the first sentence. As discussed in earlier PRs, we can assume the knowledge of high school / college maths in this spec.
  2. "If the rank of `a` is greater than 2, all except the minor 2 dimensions of `a` are treated as batch dimensions" is okay for building intuition, but it's kind of already there thanks to the first sentence.
  3. In addition to intuition, we need formal notation, to be consistent with how we specced other ops.
  4. The current version of the spec seems to have missed the "Output data is returned in the same triangle; the values in the other triangle are implementation-defined and may be anything" part of https://www.tensorflow.org/xla/operation_semantics#cholesky.

Reply:
Thanks for clearing this up.

Reply:
This is the XLA's implementation: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/cholesky_expander.cc

For a non-Hermitian matrix:
`a = [[1, 2], [3, 4]]`
The result is:
`result = [[1, 2], [0, -nan]]`

We are getting some deterministic interpreted calculation of this algorithm (Specifically Cholesky–Banachiewicz as noted in the comments to the link above). Though the result is not defined/interpretable for a non-Hermitian positive definite matrix, it is still giving an implementation defined result. Wdyt? @burmako

Reply:
"it is still giving an implementation defined result". We had an actionable conversation with Mehdi about the meaning of "implementation-defined behavior" and "undefined behavior" in https://github.com/openxla/stablehlo/pull/344#discussion_r1002609416.

In a nutshell, based on this conversation, I think that we should prefer UB over implementation-defined behavior (I'll refer to it as IB in this comment, but this is not a standard acronym) whenever we can. 

UB means that we are disallowing certain classes of programs, whereas IB means that we are allowing these programs but aren't taking a stance on their behavior. IB is very hard or impossible to walk back - once there are StableHLO programs in the wild, we cannot do much about them. As a result, if both UB and IB are an option, and we're not sure which one to pick, we should pick UB.

Now, if some behavior is legitimately allowed in StableHLO programs but may be different across various implementations, then UB is not an option, and IB might be an appropriate solution. E.g. back then, we decided that precision for many floating-point ops is implementation-defined - that's an appropriate way of using IB (UB doesn't even apply here, so it's not an option). Similarly, overflow for many integer ops is implementation-defined - that's also IB, because this is something that a lot of existing programs rely on, so we cannot make that UB.

As a result, here's the recommended rewording of the subsequent paragraph.

```
If there exists `i` where the input matrix is not an Hermitian
positive-definite matrix, then the behavior is undefined.
```